CLUSTER.md

```diff
@@ -0,0 +1,450 @@
+- Lightweight: Topology Director and Failover Coordinator located on a subset of data nodes. This removes the need to maintain and orchestrate
+non-data nodes at the cost of potentially impacting performance on the dual-use nodes.
+- Naive: Topology Director and Failover Coordinator located on all nodes. This allows for the quickest bootstrapping, but severely limits scalability.
+
+![Flotilla High Level overview](images/high_level.jpg?raw=true "High Level Overview")
+
+### The Topology Director
+
+The Topology Director serves as the source of truth for the cluster topology (excluding primary/replica status) and helps propagate
+information from Failover Coordinators. It never handles any user data, only topology metadata.
+Since it owns the metadata of the entire cluster it can allow changing this metadata while preserving the strong
+consistency guarantees we desire.
+For example - it can ensure that if a Shard holds a slot marked as Migrating, there's a corresponding Shard with the slot marked as Importing.
+We also rely on the Failover Coordinator to distribute the primary/replica state of individual nodes - This state is received periodically
+from every Failover Coordinator, cached in the Topology Director and relayed to all other Failover Coordinators when they send their state.
+
+- A completely reactive entity
+- Receives requests to alter the cluster topology and adjusts the internal state as needed
+- Validates change requests for correctness
+- Responds to topology queries by the other components
+
+#### Interfaces
+
+The Topology Director interface is fully declarative other than join/remove flows. This allows for easy orchestration via various tools.
+
+With the administrator, directly for creating the cluster / updating and removing nodes:
+
+**FLOTILLA.TOPOLOGY**: Similar to CLUSTER.SLOTS/NODES - response lists all the Flotilla nodes and their metadata  
+**FLOTILLA.REMOVE_NODE**: Remove a given node from the cluster  
+**FLOTILLA.ASSIGN_SLOTS**: Assign a set slots to specified Shard  
+**FLOTILLA.UPDATE_NODES**: Change the topology of a list of nodes    
+
+    Update/Assign_slots commands can specify the epoch they are based on in order to allow flotilla to reject outdated requests.
+    Update allows:
+    - Taking ownership of unassigned slots
+    - Moving an unallocated node / shard between Failover Coordinators
+  
+With the administrator via the individual node for adding nodes / Failover Coordinators to Flotilla:
+
+**FLOTILLA.CREATE_TD**: Create the Flotilla Cluster from the current node  
+**FLOTILLA.CREATE_FC**: Create a new Failover Coordinator entity from the current node  
+**FLOTILLA.JOIN_TD**: Add the node to be part of the Topology Director quorum  
+**FLOTILLA.JOIN_FC**: Add the node to be part of a given Failover Coordinator  
+**FLOTILLA.JOIN_NODE**: Add the data node to Flotilla  
```

Comment:
Any node joining the cluster needs to know the address of the TD(s) in order to send it the initial join request (and receives that info as one of the parameters of the join) - it doesn't need to be part of the node configuration if that's what you mean.

Naming wise - To me join implies more locality then add (as in -- I'm joining *this* entity to the cluster, as opposed to potentially adding an arbitrary one to a group) but I can consult with some tech writers on this.

Reply:
I think the question is rather about which node initiates the communication. Does the new node send a request the TD or does the TD send a request to the new node?

The word "join" as in "the new node joins the cluster" is more common than the reverse "I'm joining this entity to the cluster". The latter is kind of weird. If the latter is what it means, then I think "add" is a better word.

Reply:
> I think the question is rather about which node initiates the communication. Does the new node send a request the TD or does the TD send a request to the new node?

This is what I was thinking about. If the system has an external control plane already, it's actually inconvenient to send a command to a node to add itself. It would be more straight forward to add the API on the TD. We can, and should, still add a "JOIN CLUSTER" like API that you call on a node with a TD endpoint.

Reply:
It's possible to add such a bootstrapping mode,  it would probably only be used by the 'big' vendors and much less by the larger community.

Reply:
I think ADD vs JOIN is essentially the same as the push vs pull question.
With the current approach, it seems like there's agreement that JOIN makes for a better term.

Reply:
I just saw your discussion about this. I think we could name the interface  when we have much more detail about the interface, such as return type,  caller and function incoming parameter of the function. Maybe too much detail is beyond this spec.